### ISSUE TITLE
Accessing default component_function of lightline.vim

### ISSUE THREAD

**Issue by yous** [itchyny/lightline.vim#120]
To implement `component_function`s which are named to `MyFileformat`, `MyFiletype`, `MyFileencoding` in order to hide `fileformat`, `filetype`, `fileencoding` on narrow window respectively, we have to do like:

``` vimrc
function! MyFileformat()
  return winwidth(0) > 70 ? &fileformat : ''
endfunction

function! MyFiletype()
  return winwidth(0) > 70 ? (strlen(&filetype) ? &filetype : 'no ft') : ''
endfunction

function! MyFileencoding()
  return winwidth(0) > 70 ? (strlen(&fenc) ? &fenc : &enc) : ''
endfunction
```

But writing `&fileformat`, `(strlen(&filetype) ? &filetype : 'no ft')`, `(strlen(&fenc) ? &fenc : &enc)` things require finding the default implementations and are hard to be updated when the default implementations of lightline.vim have been changed. So I want to replace them with specific variables or functions from lightline.vim e.g. `lightline#default_component('filetype')`. Will this be possible?

**Comment by itchyny** [comment]
- No, the default implementation of lightline has not been changed for months and will never.
- It's theoretically impossible to provide the result of `component`s of lightline for users'  `component_function`s. This is because the default `component`s are just parts of statusline fields. For example, the default `readonly` component of lightline is defined by `%R`, whose actual value cannot be provided by some functions.

**Comment by yous** [comment]
Ah, I didn't noticed options like `%R`. Then it's possible to set `'component': { 'filetype': MyFiletype() }` where `MyFiletype()` returns `'%{strlen(&filetype)?&filetype:"no ft"}'`? If it's possible, then the default `lightline.component` will be useful which are defined in `lightline#init()`.

**Comment by itchyny** [comment]
Possible but I'm not sure how to extend configuration from that. You'll parse the text of the components?

**Comment by yous** [comment]
Basically, I want to use simple wrapper functions for some components like:

``` vim
function! MyMode()
  let fname = expand('%:t')
  return fname == '__Tagbar__' ? 'Tagbar' :
        \ fname == 'ControlP' ? 'CtrlP' :
        \ fname == '__Gundo__' ? 'Gundo' :
        \ fname == '__Gundo_Preview__' ? 'Gundo Preview' :
        \ fname =~ 'NERD_tree' ? 'NERDTree' :
        \ &ft == 'unite' ? 'Unite' :
        \ &ft == 'vimfiler' ? 'VimFiler' :
        \ &ft == 'vimshell' ? 'VimShell' :
        \ winwidth(0) > 60 ? lightline#mode() : ''
endfunction
```

as introduced in documentation, and this uses `lightline#mode()` as fallback. I want these for `fileformat`, `filetype`, `fileencoding`, etc. But we can't use the results like 'unix', 'vim', 'utf-8' as-is, so we should be able to get them in format of statusline.

Say the function `lightline#fileformat_statusline()` for `fileformat`, then it will return a string `'&fileformat'`. Also my vimrc will be:

``` vim
let g:lightline = {
      \ 'component': {
      \   'fileformat': LightlineFileformat() } }

function! LightlineVisible()
  let fname = expand('%:t')
  return !(fname == '__Tag_List__' ||
        \ fname == 'ControlP' ||
        \ fname =~ 'NERD_tree' ||
        \ winwidth(0) <= 70)
endfunction

function! LightlineFileformat()
  return LightlineVisible() ? lightline#fileformat_statusline() : ''
endfunction
```

As seeing `lightline#init()`, it seems that there are 20 default `component`s in lightline.vim:

``` vim
s:lightline.component = {
      \ 'mode': '%{lightline#mode()}',
      \ 'absolutepath': '%F',
      \ 'relativepath': '%f',
      \ 'filename': '%t',
      \ 'modified': '%M',
      \ 'bufnum': '%n',
      \ 'spell': '%{&spell?&spelllang:"no spell"}',
      \ 'paste': '%{&paste?"PASTE":""}',
      \ 'readonly': '%R',
      \ 'charvalue': '%b',
      \ 'charvaluehex': '%B',
      \ 'fileencoding': '%{strlen(&fenc)?&fenc:&enc}',
      \ 'fileformat': '%{&fileformat}',
      \ 'filetype': '%{strlen(&filetype)?&filetype:"no ft"}',
      \ 'percent': '%3p%%',
      \ 'percentwin': '%P',
      \ 'lineinfo': '%3l:%-2v',
      \ 'line': '%l',
      \ 'column': '%c',
      \ 'close': '%999X X ' }
```

**Comment by yous** [comment]
While writing the above comment, I think the status line won't be updated if the window size has been changed, since it's `component`, not `component_function`. Then this approach might be wrong.

**Comment by itchyny** [comment]
Yes. In the following configuration, the value of `LightlineFileformat()` is evaluated when the configuration is loaded, and will not be updated.
```vim
let g:lightline = {
      \ 'component': {
      \   'fileformat': LightlineFileformat() } }
```
Sorry but I cannot provide the feature you might want.

**Comment by yous** [comment]
Hi, I found that `component_expand` would do this. It seems to work with this vimrc:

``` vim
let g:lightline = {
      \ 'component_expand': {
      \   'percent': 'LightlinePercent',
      \   'lineinfo': 'LightlineLineinfo' } }

function! LightlinePercent()
  return LightlineVisible() ? '%3p%%' : ''
endfunction

function! LightlineLineinfo()
  return LightlineVisible() ? '%3l:%-2v' : ''
endfunction
```

I think it updates automatically without calling any explicit function of lightline.vim. What do you think?

**Comment by itchyny** [comment]
The components in `component_expand` is evaluated only once. You have a condition `winwidth(0) <= 70` but it is not applied when you narrow the window by `<C-w><`. In order to update the components in `component_expand`, you have to `call lightline#update()`.

**Comment by yous** [comment]
Oh, I see. This is going to be hard to achieve. Is there any way to manage the visibility of components depending on `winwidth(0)` easily without defining extra functions for every component? Maybe a separate issue should be made.

**Comment by itchyny** [comment]
Impossible. Sorry but I'll never implement a feature like that. You can easily control the visibility inside your component functions. Consider how to provide the feature like that. Some may say that they want to specify different threshold values for the visibility of components, other may want to specify the filetypes to control the components. Some may change the threshold depending on the filetype of the buffer. Do I have to implement and maintain such a difficult api? The easiest way to let the users control the components freely is making the plugin configurable by functions; the idea of component functions.

**Comment by yous** [comment]
Okay then, I'll try to write my own function for each component. Also it'll be very helpful if you provide documents for getting default value of components to replace flags like `%F`, `%P` used in statusline. I found `(100 * line('.') / line('$')) . '%'` for `percent` and `printf("%3d:%-2d", line('.'), col('.'))` for `lineinfo` in #112, but if there are some documentation or references in help file, it'll be great.

**Comment by yous** [comment]
Actually I want to implement [automatic truncation](https://github.com/bling/vim-airline#automatic-truncation) of statusline, so if there is a way to configure `lightline.active` and `lightline.inactive` dynamically, `component_function` for each component will be much simpler. For now, we have to define a function which checks `winwidth(0)` and the buffer is of plugin or not. If we can configure multiple or dynamic `lightline.active` and `lightline.inactive`, then the function won't have to check `winwidth(0)`.

Please tell me what do you think. If it's hard to implement or not good for design, I'll just handle them in each `component_function`.

**Comment by itchyny** [comment]
I do not understand what you mean by *dynamically*. What kind of api do you want? Can you paste some example of configurations? I will not provide apis which can be fulfilled by component functions.

**Comment by yous** [comment]
As I think, I should implement basic features which are implemented in vim-airline. I'd like to share to you when I'm done, maybe we should provide more examples for some features widely used.
